### PR TITLE
lets use sshslaves plugin instead of plain ssh machinery

### PR DESCRIPTION
**Pull request by mavlyutov**
Configuration page now looks like this: http://ibin.co/1qpuFohk4MYW (Note the Field-Group labeled "Jenkins Credentials" - this is reused from the global jenkins credentials)
Here is a sample slave log: http://ibin.co/1qpvMVMIYmof

many thanks to @felfert

### CODE REVIEW

**Comment by mavlyutov** [comment]
can anyone point me why the hell it fails with ```java.lang.NoSuchMethodError: com.google.inject.Binder.bindListener```?

**Comment by felfert** [comment]
@mavlyutov commented:
> can anyone point me why the hell it fails with java.lang.NoSuchMethodError:
> com.google.inject.Binder.bindListener?

Apparently, it is a broken secondary dependency in the newly introduced
dependeny on the ssh-slaves plugin, because:

If I add the the following into my (otherwise perfectly working)
jclouds-plugin/pom.xml, I get exactly the same error:
```text
+ <dependency>
+ <groupId>org.jenkins-ci.plugins</groupId>
+ <artifactId>ssh-slaves</artifactId>
+ <version>1.9</version>
+ </dependency>
```

So my *guess* is, that the ssh-slaves need another dependency
explicitely which has to be added in *this* pom as well...

Cheers
 -Fritz

**Comment by mavlyutov** [comment]
summon @abayer

**Comment by mavlyutov** [comment]
as ours PR is backward incompatible, we should add that information in release notes and put a warning message in update center

**Comment by felfert** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

On 02/09/2015 04:18 PM, Marat Mavlyutov wrote:
> as ours PR is backward incompatible, we should add that information in
> release notes and put a warning message in update center
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/jclouds-plugin/pull/81#issuecomment-73525408>.
> 

I'm currently exploring methods for automatic migration - just give me a
day to find out. If we are lucky - we don't need to write any warning ;)

If we cannot migrate automatically, I also would probably wait with a
until *all* those changes (the transition to credentials plugin
everywhere) are done. Then we would have to write such a warning only
*once*.

Cheers
 -Fritz
-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1

iEYEARECAAYFAlTY31oACgkQboM4mAMyprC19gCgmxAMliOVQBrHxCL5ujN3GMmx
r0gAn2FEG2MzUJzVb3otd19anjxjdpsD
=FaFR
-----END PGP SIGNATURE-----

**Comment by felfert** [comment]
On 02/09/2015 04:18 PM, Marat Mavlyutov wrote:
> as ours PR is backward incompatible, we should add that information in
> release notes and put a warning message in update center
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jenkinsci/jclouds-plugin/pull/81#issuecomment-73525408>.
> 

I'm currently exploring methods for automatic migration - just give me a
day to find out. If we are lucky - we don't need to write any warning

If we cannot migrate automatically, I also would probably wait with a
until *all* those changes (the transition to credentials plugin
everywhere) are done. Then we would have to write such a warning only
*once*.

Cheers
 -Fritz

File: jclouds-plugin/src/main/resources/jenkins/plugins/jclouds/compute/JCloudsSlaveTemplate/config.jelly [hunk 1]
```diff
@@ -96,8 +96,8 @@
           <f:textarea />
         </f:entry>
 
-        <f:entry title="Jenkins User" field="jenkinsUser">
-          <f:textbox />
+        <f:entry title="${%Jenkins Credentials}" field="credentialsId">
```
Comment (mavlyutov) [inline jclouds-plugin/src/main/resources/jenkins/plugins/jclouds/compute/JCloudsSlaveTemplate/config.jelly]: ehmm.. why do we need admin user/password if we use jenkins_user?

Comment (felfert) [inline jclouds-plugin/src/main/resources/jenkins/plugins/jclouds/compute/JCloudsSlaveTemplate/config.jelly]: Good question, next one :-)

Seriously: I was about to ask you the same. I never used it myself.

-Fritz

**Comment by mavlyutov** [comment]
okay, Im gonna merge this one today
the next steps should be:
1. remove private/public keys from global configuration
2. 'merge' jenkins_user and admin_user to single logical user
3. remove InstallJDK option as it provided via ssh-slaves

**Comment by felfert** [comment]
I have identified one scenario where the old adminuser still might be
necessary:

IF NOT using a preExistingJenkins
AND NOT allowSudo
AND NOT initscript.isEmpty

Since the execution of the initscript requires root privileges, the
admin user would be necessary in that case.

Cheers
 -Fritz

**Comment by mavlyutov** [comment]
and why do one need to use another user to start slave.jar?
by using another user, I mean folowing:
- login to server
- create new user // this code is OS dependent
- create working dir for him
- chown that dir
- su new_user
- start slave.jar

I think this is too dirty. moreover I cant see any problems to start slave.jar by existing admin user

**Comment by felfert** [comment]
First and most important:
The slave.jar should *never* be started as root. The user may have sudo
rights but this is more an exception...

One example is rpm and deb packaging. Those are best done as *non*-root,
because otherwise the install-routines could write into undesireable
locations outside of fakeroot. Like overwriting /etc/passwd with an
empty file ;)

Personally, I would get rid of the whole initscript idea instead,
because all our scenarios involve cloud-init (which runs with
admin-rigts per se). The problem is:
I cannot speak for everyone.
I only have experience with OpenStack *with* cloud-init for the
low-level provisioning (HEAT/Chef/Puppet/etc involve cloud-init as a
starting-point as well in our environment).

For those, who do *not* use cloud-init, I can *not* speak.

Cheers
 -Fritz

**Comment by mavlyutov** [comment]
ye, I got the point. So I suggest to 
1. replace global key/password parameters with global_credentialsId, 
2. replace admin_user/admin_password(which is by default = global credentialsId) with admin_credentialsId which overrides global_credentialsId
3. replace jenkinsUser(which is by default = admin_credentialIsd) with jenkinsUser_credentialsId

one moment: who's privateKey should we install with checked installPrivateKey checkbox?

**Comment by nacx** [comment]
>one moment: who's privateKey should we install with checked installPrivateKey checkbox?

That checkbox was marked to upload the private key of the admin user created by the AdminAccess statement. When creating slaves that have to clone a git repository configured to use SSH, the slave might need to have the private key in place to be able to access the git repo. IIRC, that was the purpose of that flag.

>Personally, I would get rid of the whole initscript idea instead

Is the intention to remove the inis-script feature? In that case, would users be able to specify a custom script to be executed once the slave is provisioned? I'd strongly recommend to keep that, as many users rely on that feature.

**Comment by felfert** [comment]
I think, we need  to define some common Terms and preconditions first:

The global key/password parameters cannot go into a *single*
credentialsId. The standard jenkins credentials-plugin as well as the
ssh-credentials plugin can *not* combine username-password-pairs with a key.

- A UserNamePasswordCredential has exactly one username and exactly one
password.

- A SSHPrivateKeyCredential has exactly one username, a key (actually,
it can have a list of keys) and a passphrase (if the key(s) are encrypted).

So:

1. The global *cloud* credentials can be mapped to a
UserNamePasswordCredential:
  Old field names: identity, credential
  New field name: cloudCredentialsId

The global keypair can be mapped mapped to a SSHPrivateKeyCredential:
  Old field names: privateKey, publicKey
  New field name: cloudGlobalKeyId
  => Since originally there is no username associated with this keypair,
we use the cloud type+name as some dummy username.

For the users, it is more complicated:

If useExistingJenkins is false:
 A SSHPrivateKeyCredential is created which has the default username
"jenkins" and the same keypair like the global key. The Id of this is
assigned to credentialsId and preselected. The user then can change the
username of that SSHPrivateKeyCredential if he likes.

If useExistingJenkins is true:
 The user *must* select/create an arbitrary SSHPrivateKeyCredential.

=> credentialsId may never be null and it always must be of type
SSHPrivateKeyCredential. In the current implementation this is not yet
enforced.

For the adminUser:
  Old field names: vmUser, vmPassword
  New field name adminCredentialsId
 The user *can* select/create an arbitrary SSHPrivateKeyCredential *or*
UsernamePasswordCredential. If nothing is selected, then no adminUser
can be used and the initscript will be disabled.

=> adminCredentialsId is optional.

> one moment: who's privateKey should we install with checked
> installPrivateKey checkbox?

The key from credentialsId, because this is the user who needs it during
checkout from some git or similar operations.

Now for migration strategy:

I have researched the various options mentioned at
https://wiki.jenkins-ci.org/display/JENKINS/Extend+Jenkins in the
"XStream Persistence" chapter.

In order to keep backward compatibility, we must retain the old field
names. I have started implementing a static nested class in JCloudsCloud:
ConverterImpl extends XStream2.PassthruConverter<JCloudsCloud> which
would take care of the global conversions. Secifically, it creates the
objects for cloudCredentialsId and cloudGlobalKeyId from the old fields
content.
For the jenkins- and admin-user, the same has to be done in the
JCloudsSlaveTemplate.

If those two are implemented, hopefulle nobody even notices the change
when upgrading ... ;)

>
> one moment: who's privateKey should we install with checked
> installPrivateKey checkbox?

The key of credentialsId (the old jenkinsUser), because this is the user
who needs it during checkout from some git or similar operations during
build operations.

Cheers
 -Fritz

**Comment by felfert** [comment]
On 02/10/2015 02:20 PM, Ignasi Barrera wrote:
>     one moment: who's privateKey should we install with checked
>     installPrivateKey checkbox?
> 
> That checkbox was marked to upload the private key of the admin user
> created by the AdminAccess statement. When creating slaves that have to
> clone a git repository configured to use SSH, the slave might need to
> have the private key in place to be able to access the git repo. IIRC,
> that was the purpose of that flag.
> 
Thanks for clearing that.

>     Personally, I would get rid of the whole initscript idea instead
> 
> Is the intention to remove the inis-script feature? In that case, would
> users be able to specify a custom script to be executed once the slave
> is provisioned? I'd strongly recommend to keep that, as many users rely
> on that feature.
>

No, certainly not. That was just my personal opinion. (I find it
useless, because we use cloud-init here which covers everything already.)

Cheers
 -Fritz

**Comment by mavlyutov** [comment]
>1. The global *cloud* credentials can be mapped to a
>UserNamePasswordCredential:
>  Old field names: identity, credential
>  New field name: cloudCredentialsId
we don't need to hide it behind cloudCredentialsId, let's leave it as is

>The global keypair can be mapped mapped to a SSHPrivateKeyCredential:
>  Old field names: privateKey, publicKey
>  New field name: cloudGlobalKeyId
>  => Since originally there is no username associated with this keypair,
>we use the cloud type+name as some dummy username.
let's use 'root' or 'cloud-user' as dummy username - this is the most popular usecase

**Comment by felfert** [comment]
On 02/10/2015 02:52 PM, Marat Mavlyutov wrote:
>     => Since originally there is no username associated with this keypair,
>     we use the cloud type+name as some dummy username.
>     let's use 'root' or 'cloud-user' as dummy username - this is the
>     most popular usecase

Have a look at this screenshot:

http://ibin.co/1rBVVWEKxGG1

The username is completely free in jenkins credentials store. If we name
it always the same, it may look like 3rd and 4th entry in the screenshot
and you don't know to which cloud it belongs and if it is a real user or
just the global key.

Why not going for the last variant in that screenshot (Username is
"Global key", comment is "Jclouds cloud othercloud")?

Of course, all of the above applies only to automatically migrated
stuff. The user can always specify something else.

Cheers
 -Fritz

**Pull request merged**